File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.py
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        sums = [1] + [0 for i in range(k)]
        total = 0
        result = 0

        for num in nums:
            total += num
            total %= k
            if sums[total]: result += sums[total]
            sums[total] += 1

        return result

File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.java
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        int[] sums = new int[k + 1];
        int total = 0, result = 0;
        sums[0] = 1;

        for (int num: nums) {
            total += num;
            total %= k;
            if (total < 0) total += k;
            if (sums[total] != 0) result += sums[total];
            sums[total]++;
        }

        return result;
    }
}